

max 0.65x1 + 0.75x2 − 5.85x3 − 8.35x4 − 12.38x5 + 30x6,
s.t. − 0.5x1 − 3.8x3 − 0.9x4 + 0.9x5 ≤ 24.8,
x1 − x4 − x6 ≤ 0.8,
−2.9x2 − 1.9x3 − 0.9x6 ≤ −71.4,
4.7x4 + 8.02x5 + 6.15x6 ≤ 95,
−0.7x1 − 1.4x2 + 7.9x3 − 2.9x4 + 0.9x5 ≤ −47,
xi ≥ 0 ∀i ∈ {1, 2, . . . , 6},


In [ ]:
!pip install -q pyomo

In [ ]:
from pyomo.environ import * 

In [ ]:
import numpy as np

In [ ]:
coef = np.loadtxt('lab4_ex1_coef.txt' ,delimiter=',')

In [ ]:
print(coef.shape)
print('Num rows:',coef.shape[0])
print('Num cols:',coef.shape[1])



(6, 7)
Num rows: 6
Num cols: 7


In [ ]:
print(coef)

[[  0.65   0.75  -5.85  -8.35 -12.38  30.     0.  ]
 [ -0.5    0.    -3.8   -0.9    0.9    0.    24.8 ]
 [  1.     0.     0.    -1.     0.    -1.     0.8 ]
 [  0.    -2.9   -1.9    0.     0.    -0.9  -71.4 ]
 [  0.     0.     0.     4.7    8.02   6.15  95.  ]
 [ -0.7   -1.4    7.9   -2.9    0.9    0.   -47.  ]]


In [ ]:
# create a model
model_lab4 = ConcreteModel()

In [ ]:
N = coef.shape[1]-1

In [ ]:
M = coef.shape[0]-1

In [ ]:
obj_coef = coef[0,:-1]
print(obj_coef.shape)
print(obj_coef)


(6,)
[  0.65   0.75  -5.85  -8.35 -12.38  30.  ]


In [ ]:
constr_coef = coef[1:,:-1]
print(constr_coef.shape)
print(constr_coef)

(5, 6)
[[-0.5   0.   -3.8  -0.9   0.9   0.  ]
 [ 1.    0.    0.   -1.    0.   -1.  ]
 [ 0.   -2.9  -1.9   0.    0.   -0.9 ]
 [ 0.    0.    0.    4.7   8.02  6.15]
 [-0.7  -1.4   7.9  -2.9   0.9   0.  ]]


In [ ]:
constr_rhs = coef[1:,-1]
print(constr_rhs.shape)
print(constr_rhs)

(5,)
[ 24.8   0.8 -71.4  95.  -47. ]


In [ ]:
lower_bound = 0
upper_bound = np.inf

In [ ]:
col_indices = np.arange(N)
print(col_indices)

[0 1 2 3 4 5]


In [ ]:
model_lab4.x = Var(col_indices)

In [ ]:
row_indices = np.arange(M)
print(row_indices)

[0 1 2 3 4]


In [ ]:
model_lab4.constraints = ConstraintList()

In [ ]:
for i in row_indices:
  model_lab4.constraints.add(summation(constr_coef[i],model_lab4.x) <= constr_rhs[i])

In [ ]:
for j in col_indices:
  model_lab4.x[j].setlb(lower_bound)
  model_lab4.x[j].setub(upper_bound)

In [ ]:
model_lab4.objective = Objective(expr = summation(obj_coef,model_lab4.x), sense=maximize)

In [ ]:
model_lab4.pprint()

2 Set Declarations
    constraints_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    5 : {1, 2, 3, 4, 5}
    x_index : Size=1, Index=None, Ordered=False
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    6 : {0, 1, 2, 3, 4, 5}

1 Var Declarations
    x : Size=6, Index=x_index
        Key : Lower : Value : Upper : Fixed : Stale : Domain
          0 :     0 :  None :  None : False :  True :  Reals
          1 :     0 :  None :  None : False :  True :  Reals
          2 :     0 :  None :  None : False :  True :  Reals
          3 :     0 :  None :  None : False :  True :  Reals
          4 :     0 :  None :  None : False :  True :  Reals
          5 :     0 :  None :  None : False :  True :  Reals

1 Objective Declarations
    objective : Size=1, Index=None, Active=True
        Key  : Active : Sense    : Expression
        None :   True : maximize : 0.65*x[0] + 0.75*x[1] - 5.85*x[2

In [ ]:
!apt-get install -y -qq coinor-cbc


In [ ]:
opt_cbc = SolverFactory('cbc')

In [ ]:
result = opt_cbc.solve(model_lab4)
print('Solver status:', result.solver.status)
print('Solver termination condition:',result.solver.termination_condition)

  - termination condition: unbounded
  - message from solver: <undefined>


Solver status: warning
Solver termination condition: unbounded


solver status and  termination condition is wrning and unbounded,
beaucse the upper bound if inf   then cosntrint and objective go to inf  then Solver termination condition: unbounded
variable are non negetive upper bound are not bounded then constraint are not  satify.


now chnages to sense= minimize

In [ ]:
model_lab4.objective.set_sense(minimize)

In [ ]:
model_lab4.pprint()

2 Set Declarations
    constraints_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    5 : {1, 2, 3, 4, 5}
    x_index : Size=1, Index=None, Ordered=False
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    6 : {0, 1, 2, 3, 4, 5}

1 Var Declarations
    x : Size=6, Index=x_index
        Key : Lower : Value : Upper : Fixed : Stale : Domain
          0 :     0 :  None :  None : False :  True :  Reals
          1 :     0 :  None :  None : False :  True :  Reals
          2 :     0 :  None :  None : False :  True :  Reals
          3 :     0 :  None :  None : False :  True :  Reals
          4 :     0 :  None :  None : False :  True :  Reals
          5 :     0 :  None :  None : False :  True :  Reals

1 Objective Declarations
    objective : Size=1, Index=None, Active=True
        Key  : Active : Sense    : Expression
        None :   True : minimize : 0.65*x[0] + 0.75*x[1] - 5.85*x[2

In [ ]:
result = opt_cbc.solve(model_lab4)
print('Solver status:', result.solver.status)
print('Solver termination condition:',result.solver.termination_condition)

  - termination condition: unbounded
  - message from solver: <undefined>


Solver status: warning
Solver termination condition: unbounded


is unbounded becuase upper bounded is inf now if upper bound is finite in this case cheek

In [ ]:
model_lab4.objective.set_sense(maximize)

In [ ]:
lower_bound_1=0
upper_bound_1=500

In [ ]:
for j in col_indices:
  model_lab4.x[j].setlb(lower_bound_1)
  model_lab4.x[j].setub(upper_bound_1)

In [ ]:
model_lab4.pprint()

2 Set Declarations
    constraints_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    5 : {1, 2, 3, 4, 5}
    x_index : Size=1, Index=None, Ordered=False
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    6 : {0, 1, 2, 3, 4, 5}

1 Var Declarations
    x : Size=6, Index=x_index
        Key : Lower : Value : Upper : Fixed : Stale : Domain
          0 :     0 :  None :   500 : False :  True :  Reals
          1 :     0 :  None :   500 : False :  True :  Reals
          2 :     0 :  None :   500 : False :  True :  Reals
          3 :     0 :  None :   500 : False :  True :  Reals
          4 :     0 :  None :   500 : False :  True :  Reals
          5 :     0 :  None :   500 : False :  True :  Reals

1 Objective Declarations
    objective : Size=1, Index=None, Active=True
        Key  : Active : Sense    : Expression
        None :   True : maximize : 0.65*x[0] + 0.75*x[1] - 5.85*x[2

In [ ]:
result = opt_cbc.solve(model_lab4)
print('Solver status:', result.solver.status)
print('Solver termination condition:',result.solver.termination_condition)

Solver status: ok
Solver termination condition: optimal


In [ ]:
# display solution
print('\nObjective = ', model_lab4.objective())

print('\nDecision Variables')
for i in col_indices:
  print('x[',i,'] = ', model_lab4.x[i].value)

print('\nConstraints')
model_lab4.constraints.display()


Objective =  848.9752701

Decision Variables
x[ 0 ] =  16.247154
x[ 1 ] =  500.0
x[ 2 ] =  0.0
x[ 3 ] =  0.0
x[ 4 ] =  0.0
x[ 5 ] =  15.447154

Constraints
constraints : Size=5
    Key : Lower : Body               : Upper
      1 :  None :          -8.123577 :  24.8
      2 :  None : 0.7999999999999989 :   0.8
      3 :  None :      -1463.9024386 : -71.4
      4 :  None :         94.9999971 :  95.0
      5 :  None :       -711.3730078 : -47.0


AttributeError: ignored

upper bound is bound then optimal solution are exits, constraint are satifys this then optimal solution exit
bjective =  848.9752701

Decision Variables
x[ 0 ] =  16.247154
x[ 1 ] =  500.0
x[ 2 ] =  0.0
x[ 3 ] =  0.0
x[ 4 ] =  0.0
x[ 5 ] =  15.447154
 constraint 2 and constraint 4 are active 